### PR TITLE
GEOMESA-2552 Redis data store implementation

### PR DESCRIPTION
**Pull request by elahrvivaz**
* Each index is a Redis SortedSet, with the values comprised of
  concatenated index keys and row values
* Change default query planning to heuristic based

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

File: build/cqs.tsv [hunk 1]
```diff
@@ -237,7 +237,7 @@ org.scalatra:scalatra-swagger_2.11	2.6.3	compile
 org.slf4j:slf4j-api	1.7.25	compile
 org.slf4j:slf4j-log4j12	1.7.25	compile
 picocontainer:picocontainer	1.2	compile
-redis.clients:jedis	2.9.0	compile
+redis.clients:jedis	3.0.1	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: approved: http://dev.eclipse.org/ipzilla/show_bug.cgi?id=18915

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseDataStore.scala [hunk 2]
```diff
@@ -48,23 +49,12 @@ class HBaseDataStore(val connection: Connection, override val config: HBaseDataS
...
 
-  override val stats: GeoMesaStats =
-    if (config.remoteFilter) { new DistributedRunnableStats(this) } else { new UnoptimizedRunnableStats(this) }
+  override val stats: GeoMesaStats = new RunnableStats(this)
 
   // zookeeper locking
   override protected val mock: Boolean = false
   override protected val zookeepers: String = ZKConfig.getZKQuorumServersString(connection.getConfiguration)
 
-  override def delete(): Unit = {
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseDataStore.scala]: this is implemented already in the super class, and this method doesn't do anything different so was removed

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala [hunk 3]
```diff
@@ -55,7 +55,6 @@ object HBaseQueryPlan {
   def explain(plan: HBaseQueryPlan, explainer: Explainer, prefix: String): Unit = {
     explainer.pushLevel(s"${prefix}Plan: ${plan.getClass.getSimpleName}")
     explainer(s"Tables: ${plan.tables.mkString(", ")}")
-    explainer(s"Filter: ${plan.filter.toString}")
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: this was a redundant message as the filter is already logged by the query planner just before calling this method

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/attribute/AttributeIndex.scala [hunk 4]
```diff
@@ -80,28 +56,6 @@ class AttributeIndex protected (ds: GeoMesaDataStore[_],
       opt
     }
   }
-
-  // note: needs to be lazy to allow for keyspace override in subclasses
-  protected lazy val rowValueOffset: Int = keySpace.sharding.length
```
Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/attribute/AttributeIndex.scala]: the logic removed from this class was moved in the keyspace instead

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/id/legacy/IdIndexV3.scala [hunk 5]
```diff
@@ -36,17 +36,9 @@ object IdIndexV3 {
 
   class IdIndexKeySpaceV3(sft: SimpleFeatureType, override val sharing: Array[Byte]) extends IdIndexKeySpace(sft) {
 
-    private val rangePrefixes = {
```
Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/id/legacy/IdIndexV3.scala]: shards are not used by the id index, so this block was redundant as sharding.length will always be 0

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/metadata/MetadataSerializer.scala [hunk 6]
```diff
@@ -11,15 +11,15 @@ package org.locationtech.geomesa.index.metadata
 import java.nio.charset.StandardCharsets
 
 trait MetadataSerializer[T] {
-  def serialize(typeName: String, key: String, value: T): Array[Byte]
-  def deserialize(typeName: String, key: String, value: Array[Byte]): T
+  def serialize(typeName: String, value: T): Array[Byte]
```
Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/metadata/MetadataSerializer.scala]: `key` was not actually used in any of these methods, so I removed it from the signature.

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/stats/MetadataBackedStats.scala [hunk 7]
```diff
@@ -136,14 +140,41 @@ trait MetadataBackedStats extends GeoMesaStats with StatsBasedEstimator with Laz
...
+      case hm: HasGeoMesaMetadata[String] =>
+        val date = GeoToolsDateFormat.format(Instant.now().atZone(ZoneOffset.UTC))
+        hm.metadata.insert(sft.getTypeName, GeoMesaMetadata.STATS_GENERATION_KEY, date)
+      case _ => // no-op
+    }
 
     stats
   }
 
+  override def runStats[T <: Stat](sft: SimpleFeatureType, stats: String, filter: Filter): Seq[T] = {
```
Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/stats/MetadataBackedStats.scala]: previously this was delegated to individual data stores, but at this point all data stores can handle a 'stats' type query

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/stats/package.scala [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+ ***********************************************************************/
+
+package org.locationtech.geomesa.index
+
+package object stats {
+
+  /**
+    * Marker trait for classes with stats
+    */
+  trait HasGeoMesaStats {
```
Comment (elahrvivaz) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/stats/package.scala]: moved this class to a package object instead of having multiple top-level classes in a single file

File: pom.xml [hunk 9]
```diff
@@ -300,916 +301,321 @@
 
     <dependencyManagement>
         <dependencies>
+            <!-- geomesa modules -->
```
Comment (elahrvivaz) [inline pom.xml]: I re-organized the pom to group things together better, but the imports did not change except for the addition of jedis

File: docs/tutorials/geomesa-examples-avro.rst [hunk 10]
```diff
@@ -17,7 +17,7 @@ Choice of Backing Storage
 -------------------------
 
 This tutorial will work with several different back-ends. For simplicity, the rest of the tutorial will assume
-the use of HBase. Alternatively, you may use Accumulo, Cassandra, or the GeoMesa FileSystem DataStore. If not
+the use of HBase. Alternatively, you may use Accumulo, Cassandra, Redis,, or the GeoMesa FileSystem DataStore. If not
```
Comment (jnh5y) [inline docs/tutorials/geomesa-examples-avro.rst]: s/,,/,/g

File: docs/user/redis/install.rst [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+classpath or add the file to a zip or JAR archive to be included on the classpath.
+
+.. _install_redis_geoserver:
+
+Installing GeoMesa Redis in GeoServer
+-------------------------------------
+
+.. warning::
+
+    GeoMesa 2.2.0 and later require GeoServer 2.14.0 or later. GeoMesa 2.1.0 and earlier require GeoServer 2.12.5.
```
Comment (jnh5y) [inline docs/user/redis/install.rst]: I know this is boilerplate.  What's the story with GeoServer 2.13 again?  2.13.0 and 2.13.1 had some issues.  What about 2.13.2+?

File: docs/user/upgrade.rst [hunk 12]
```diff
@@ -90,6 +90,16 @@ Compatibility Matrix
...
 
+Version 2.3.0 Upgrade Guide
++++++++++++++++++++++++++++
+
+Default Query Planning Type
+---------------------------
+
+GeoMesa 2.3.0 changes the default query planning type from stat-based to heuristic-based. This will only affect the
+Accumulo data store, as other stores have not implemented statistics. To enable stat-based query planning, refer
+to :ref:`query_planning_hint`.
```
Comment (jnh5y) [inline docs/user/upgrade.rst]: With this default... does GeoMesa on Accumulo generate the stats tables by default?  Or does one have to opt into that?

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/accumulo/data/AccumuloDataStoreColumnGroupsTest.scala [hunk 13]
```diff
@@ -101,7 +101,7 @@ class AccumuloDataStoreColumnGroupsTest extends Specification with TestWithDataS
     ("bbox(geom,45.4,54.9,45.8,55.1) and height < 67", 4 to 6),
     ("age >= 50 and height < 67", 5 to 6),
     ("bbox(geom,45.4,54.9,45.8,55.1) AND name IN ('name5', 'name6') and height < 67", 5 to 6),
-    ("bbox(geom,45.4,54.9,45.8,55.1) AND age >= 50 and height < 67", 5 to 6),
+    ("bbox(geom,45.41,54.9,45.8,55.1) AND age IS NOT NULL and height < 67", 5 to 6),
```
Comment (jnh5y) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/accumulo/data/AccumuloDataStoreColumnGroupsTest.scala]: Why the change in the filter here?

File: docs/user/redis/install.rst (hunk 11 above)
Comment (elahrvivaz) [reply docs/user/redis/install.rst]: I think 2.13.2+ is ok, but we haven't really run anything on it so I figured it was safer to recommend the tried and tested 2.12 version.

File: docs/user/upgrade.rst (hunk 12 above)
Comment (elahrvivaz) [reply docs/user/upgrade.rst]: it will generate it by default, but you can turn it off in the data store params or via sys property.

Comment (elahrvivaz) [reply docs/user/upgrade.rst]: so it will still track stats, just won't use them for query planning. But they are available for bounds, counts, etc

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/accumulo/data/AccumuloDataStoreColumnGroupsTest.scala (hunk 13 above)
Comment (elahrvivaz) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/accumulo/data/AccumuloDataStoreColumnGroupsTest.scala]: with the change to heuristic query planning, the filter was hitting a different index, so i tweaked it to keep it against the z3

**Comment by jnh5y** [comment]
Needs to add Redis info to the root README.md

File: docs/tutorials/geomesa-quickstart-redis.rst [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+``export`` command of the GeoMesa Redis tools distribution. For a more production ready example, you
+can alternatively stand up a GeoServer and connect it to your Redis instance.
+
+Visualize Data With Leaflet
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. warning::
+
+    To successfully run this command you must have a computer that is connected to the internet
+    in order to access external leaflet resources.
```
Comment (jnh5y) [inline docs/tutorials/geomesa-quickstart-redis.rst]: Capitalize Leaflet.

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/api/IndexKeySpace.scala [hunk 15]
```diff
@@ -37,11 +37,12 @@ trait IndexKeySpace[T, U] {
...
 
   /**
-    * Length of an index key
+    * Length of an index key. If static (general case), will return a Right with the length. If dynamic,
+    * will return Left with a function to determine the length from a given (row, offset, length)
     *
     * @return
     */
-  def indexKeyByteLength: Int
+  def indexKeyByteLength: Either[(Array[Byte], Int, Int) => Int, Int]
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/api/IndexKeySpace.scala]: *blinks*  Ok, which situations require the dynamic approach?

File: geomesa-redis/geomesa-redis-datastore/src/main/scala/org/locationtech/geomesa/redis/data/RedisDataStore.scala [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+
+  @throws(classOf[IllegalArgumentException])
+  override protected def preSchemaCreate(sft: SimpleFeatureType): Unit = {
+    if (sft.getVisibilityLevel == VisibilityLevel.Attribute) {
+      throw new IllegalArgumentException("Attribute level visibility is not supported in this store")
+    }
+
+    sft.getAttributeDescriptors.asScala.foreach { descriptor =>
+      if (descriptor.getColumnGroups().nonEmpty) {
+        throw new IllegalArgumentException("Column groups are not supported in this store")
```
Comment (jnh5y) [inline geomesa-redis/geomesa-redis-datastore/src/main/scala/org/locationtech/geomesa/redis/data/RedisDataStore.scala]: Do all stores throw exceptions like this for unsupported sft options?  

I wonder if it'd be worth taking an approach like ContentDataStore where an implementing class sets various things to true or false to indicate which things are supported/unsupported.

File: geomesa-redis/geomesa-redis-datastore/src/main/scala/org/locationtech/geomesa/redis/data/RedisDataStore.scala [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+      }
+    }
+
+    // disable shards
+    sft.setZShards(0)
+    sft.setIdShards(0)
+    sft.setAttributeShards(0)
```
Comment (jnh5y) [inline geomesa-redis/geomesa-redis-datastore/src/main/scala/org/locationtech/geomesa/redis/data/RedisDataStore.scala]: Interesting...  I wonder if sharding + Redis clustering would make any sense...

File: docs/tutorials/geomesa-quickstart-redis.rst (hunk 14 above)
Comment (elahrvivaz) [reply docs/tutorials/geomesa-quickstart-redis.rst]: fixed, thanks

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/api/IndexKeySpace.scala (hunk 15 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/api/IndexKeySpace.scala]: just the attribute index at the moment...

File: geomesa-redis/geomesa-redis-datastore/src/main/scala/org/locationtech/geomesa/redis/data/RedisDataStore.scala (hunk 16 above)
Comment (elahrvivaz) [reply geomesa-redis/geomesa-redis-datastore/src/main/scala/org/locationtech/geomesa/redis/data/RedisDataStore.scala]: that's not a bad idea, although i'd prefer it to be declarative instead of knowing that you have to override something. I opened a ticket here: https://geomesa.atlassian.net/browse/GEOMESA-2565

File: geomesa-redis/geomesa-redis-datastore/src/main/scala/org/locationtech/geomesa/redis/data/RedisDataStore.scala (hunk 17 above)
Comment (elahrvivaz) [reply geomesa-redis/geomesa-redis-datastore/src/main/scala/org/locationtech/geomesa/redis/data/RedisDataStore.scala]: Clustering can help with having more indices, but since each index 'table' is a single key, it will live on a single redis node regardless...

File: docs/tutorials/geomesa-examples-avro.rst (hunk 10 above)
Comment (elahrvivaz) [reply docs/tutorials/geomesa-examples-avro.rst]: fixed, thanks
